casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -20,54 +22,72 @@ object ProtoUtil {
     if (candidate == target) {
-      true
+      true.pure[F]
     } else {
-      (for {
-        hdr        <- target.header
-        parentHash <- hdr.parentsHashList.headOption
-        mainParent <- internalMap.get(parentHash)
-      } yield mainParent) match {
-        case Some(parent) => isInMainChain(internalMap, candidate, parent)
-        case None         => false
+      val maybeMainParentHash = ProtoUtil.parentHashes(target).headOption
+      maybeMainParentHash match {
+        case Some(mainParentHash) =>
+          for {
+            mainParent <- BlockStore[F].get(mainParentHash)
+            isInMainChain <- mainParent match {
+                              case Some(parent) => isInMainChain[F](candidate, parent)
+                              case None         => false.pure[F]
+                            }
+          } yield isInMainChain
+        case None => false.pure[F]
       }
     }
 
-  @tailrec
-  def getMainChain(internalMap: Map[BlockHash, BlockMessage],
-                   estimate: BlockMessage,
-                   acc: IndexedSeq[BlockMessage]): IndexedSeq[BlockMessage] = {
-    val parentsHashes       = ProtoUtil.parents(estimate)
+  def getMainChain[F[_]: Monad: BlockStore](
+      estimate: BlockMessage,
+      acc: IndexedSeq[BlockMessage]): F[IndexedSeq[BlockMessage]] = {
+    val parentsHashes       = ProtoUtil.parentHashes(estimate)
     val maybeMainParentHash = parentsHashes.headOption
-    maybeMainParentHash flatMap internalMap.get match {
-      case Some(newEstimate) =>
-        getMainChain(internalMap, newEstimate, acc :+ estimate)
-      case None => acc :+ estimate
-    }
+    for {
+      mainChain <- maybeMainParentHash match {
+                    case Some(mainParentHash) =>
+                      for {
+                        updatedEstimate <- unsafeGetBlock[F](mainParentHash)
+                        mainChain       <- getMainChain[F](updatedEstimate, acc :+ estimate)
+                      } yield mainChain
+                    case None => (acc :+ estimate).pure[F]
+                  }
+    } yield mainChain
   }
 
-  @tailrec
-  def findJustificationParentWithSeqNum(b: BlockMessage,
-                                        internalMap: Map[BlockHash, BlockMessage],
-                                        seqNum: SequenceNumber): Option[BlockMessage] =
+  def unsafeGetBlock[F[_]: Monad: BlockStore](hash: BlockHash): F[BlockMessage] =
+    for {
+      maybeBlock <- BlockStore[F].get(hash)
+      block = maybeBlock match {
+        case Some(b) => b
+        case None =>
+          throw new Error(s"BlockStore is missing hash ${PrettyPrinter.buildString(hash)}")
+      }
+    } yield block
+
+  def findJustificationParentWithSeqNum[F[_]: Monad: BlockStore](
+      b: BlockMessage,
+      seqNum: SequenceNumber): F[Option[BlockMessage]] =
     if (b.seqNum == seqNum) {
-      Some(b)
+      Option[BlockMessage](b).pure[F]
```

Comment:
You could probably express this whole `findJustificationParentWithSeqNum` function slightly more elegantly with bfTraverseF and then StreamT.findF

Reply:
Done